lms/static/js/edxnotes/utils/notes_collector.js

```diff
@@ -0,0 +1,111 @@
+(function(define) {
+    'use strict';
+    define(['jquery', 'underscore', 'annotator_1.2.9'], function($, _, Annotator) {
+        var fetchOnce = null,
+            fetchNotes,
+            renderNotes,
+            _requestsData = null,
+            storeNotesRequestData,
+            SearchRequestsData = [],
+            waitUntilAllRequestsStored;
+
+        waitUntilAllRequestsStored = function (maxTimeout) {
+            var deferred = $.Deferred(),
+                intervalId = null,
+                elapsedTimeout = 0,
+                pollInterval = 100,
+                checkRequests;
+
+            checkRequests = function() {
+               if ($('.edx-notes-wrapper').length === SearchRequestsData.length) {
+                   deferred.resolve();
+               } else {
+                   elapsedTimeout += pollInterval;
+                   if (elapsedTimeout >= maxTimeout) {
+                       clearInterval(intervalId);
+                       deferred.reject();
+                   }
+               }
+            };
+
+            intervalId = setInterval(checkRequests, pollInterval);
+            return deferred.promise();
+        };
+
+        storeNotesRequestData = function (data) {
```

Comment:
I find the way this code fits together to be confusing, which may be in the naming, or may just be because the annotatorJS API is difficult to deal with. If I understand this correctly, each HTML component calls the factory via this code in ``VisibilityDecorator``:

        createNote = function(element, params) {
            if (params) {
                return NotesFactory.factory(element, params);
            }
            return null;
        };

If there are multiple HTML components, then there will be N calls to the factory. All N calls will have the same value for params, which is what is used to construct the AJAX call. However, when the user switches units, the params will likely be different, in which case a new AJAX call needs to be issued. This is also the same problem with the unit tests, which is that once an AJAX call has been issued this code will never issue another one.

It seems that this needs to be a cache, where the key is the params, and the result is a promise for the notes. If different params are passed in, then a new AJAX call needs to be issued, and a new promise returned. It looks like VisibilityDecorator also has an explicit ``cleanup`` method which will need to somehow clear the cache too.

Given these thoughts, I think this method should be called ``renderNotes`` since that is what its purpose is. The fact that it is asynchronous shouldn't matter to the caller, and the fact that it is optimizing the number of AJAX calls is also an implementation detail. It should maintain its own cache internally to track ongoing AJAX requests, and it should wait on a promise in order to render the result. Subsequent callers will wait on the same promise so only a single AJAX call will be made. This then removes all need for polling in this file.